File: app/src/main/java/com/byd/dashcast/proxy/daemon/Phase4DisplayVerbs.java
package com.byd.dashcast.proxy.daemon;

import android.os.IBinder;

import java.lang.reflect.Method;

/**
 * Phase4DisplayVerbs — display-management verbs that run inside the daemon
 * process (uid 2000).
 *
 * <p>Covers three concern groups:
 * <ol>
 *   <li><b>IWindowManager</b> — overscan insets via the {@code window} service.</li>
 *   <li><b>VirtualDisplay</b> — create / release VDs on behalf of the app (Phase 5).</li>
 *   <li><b>ATM display policy</b> — {@code setDisplayToSingleTaskInstance},
 *       {@code isActivityStartAllowedOnDisplay}.</li>
 * </ol>
 *
 * @see Phase4ProcessVerbs
 * @see Phase4TaskVerbs
 * @since v1.1.9 build 174 — split from Phase4Verbs in v1.4.4-beta.
 */
public final class Phase4DisplayVerbs {

    private Phase4DisplayVerbs() {}

    // ─── IWindowManager cache ─────────────────────────────────────────────

    private static volatile Object  sWindowManager;
    private static volatile IBinder sWindowManagerBinder;
    private static volatile Method  sSetOverscan;

    private static final IBinder.DeathRecipient sWindowManagerDeath = new IBinder.DeathRecipient() {
        @Override public void binderDied() {
            synchronized (Phase4DisplayVerbs.class) {
                sWindowManager       = null;
                sWindowManagerBinder = null;
            }
        }
    };

    private static Object windowManager() throws Throwable {
        IBinder cached = sWindowManagerBinder;
        if (cached != null && cached.isBinderAlive()) return sWindowManager;
        synchronized (Phase4DisplayVerbs.class) {
            cached = sWindowManagerBinder;
            if (cached != null && cached.isBinderAlive()) return sWindowManager;
            Class<?> sm = Class.forName("android.os.ServiceManager");
            IBinder b = (IBinder) sm.getMethod("getService", String.class).invoke(null, "window");
            if (b == null) throw new IllegalStateException("no 'window' service");
            Class<?> stub = Class.forName("android.view.IWindowManager$Stub");
            Object wm = stub.getMethod("asInterface", IBinder.class).invoke(null, b);
            if (wm == null) throw new IllegalStateException("IWindowManager.asInterface returned null");
            try { b.linkToDeath(sWindowManagerDeath, 0); }
            catch (Throwable t) { return wm; }
            sWindowManagerBinder = b;
            sWindowManager       = wm;
            return wm;
        }
    }

    private static Method setOverscanMethod() throws Throwable {
        Method m = sSetOverscan;
        if (m != null) return m;
        synchronized (Phase4DisplayVerbs.class) {
            m = sSetOverscan;
            if (m != null) return m;
            Class<?> iface = Class.forName("android.view.IWindowManager");
            m = iface.getMethod("setOverscan",
                    int.class, int.class, int.class, int.class, int.class);
            sSetOverscan = m;
            return m;
        }
    }

    // ─── VirtualDisplay map ───────────────────────────────────────────────

    private static final java.util.concurrent.ConcurrentHashMap<Integer, android.hardware.display.VirtualDisplay>
            sVirtualDisplays = new java.util.concurrent.ConcurrentHashMap<>();

    // ─── Verbs ────────────────────────────────────────────────────────────

    /**
     * Equivalent of {@code wm overscan L,T,R,B -d displayId}.
     * Probe P1 (build 173) confirmed viable from uid 2000 on BYD Seal EU.
     */
    public static void setOverscan(int displayId, int left, int top, int right, int bottom)
            throws Throwable {
        Object wm = windowManager();
        Method m = setOverscanMethod();
        m.invoke(wm, displayId, left, top, right, bottom);
    }

    /**
     * Create a VirtualDisplay on behalf of the calling app and return its display id.
     * The daemon runs as uid 2000 (shell) — uses a {@code com.android.shell} package
     * context so DisplayManagerService's uid-package check passes.
     *
     * <p>Callers MUST eventually call {@link #releaseVirtualDisplay}; Java GC does NOT
     * release the underlying IVirtualDisplay token.
     */
    public static int createVirtualDisplay(android.content.Context sysCtx,
                                           String name, int width, int height, int dpi,
                                           android.view.Surface surface, int flags) {
        if (sysCtx == null) throw new IllegalStateException("system context null");
        if (name == null || name.isEmpty()) name = "DashCast_VD";
        if (width <= 0 || height <= 0 || dpi <= 0) {
            throw new IllegalArgumentException("bad geometry " + width + "x" + height + "@" + dpi);
        }
        if (surface == null || !surface.isValid()) {
            throw new IllegalArgumentException("surface null or invalid");
        }
        android.content.Context shellCtx;
        try {
            shellCtx = sysCtx.createPackageContext("com.android.shell",
                    android.content.Context.CONTEXT_INCLUDE_CODE
                            | android.content.Context.CONTEXT_IGNORE_SECURITY);
        } catch (android.content.pm.PackageManager.NameNotFoundException nnfe) {
            throw new IllegalStateException("com.android.shell context unavailable", nnfe);
        }
        android.hardware.display.DisplayManager dm =
                (android.hardware.display.DisplayManager) shellCtx.getSystemService(
                        android.content.Context.DISPLAY_SERVICE);
        if (dm == null) throw new IllegalStateException("DisplayManager null");
        android.hardware.display.VirtualDisplay vd =
                dm.createVirtualDisplay(name, width, height, dpi, surface, flags);
        if (vd == null) throw new IllegalStateException("createVirtualDisplay returned null");
        android.view.Display d = vd.getDisplay();
        if (d == null) {
            try { vd.release(); } catch (Throwable ignore) {}
            throw new IllegalStateException("VirtualDisplay.getDisplay() null");
        }
        int id = d.getDisplayId();
        sVirtualDisplays.put(id, vd);
        return id;
    }

    /** Release a VD previously created via {@link #createVirtualDisplay}. No-op if unknown. */
    public static void releaseVirtualDisplay(int displayId) {
        android.hardware.display.VirtualDisplay vd = sVirtualDisplays.remove(displayId);
        if (vd != null) {
            try { vd.release(); } catch (Throwable ignore) {}
        }
    }

    /** Release every VD held by the daemon. Best-effort cleanup. */
    public static void releaseAllVirtualDisplays() {
        for (java.util.Map.Entry<Integer, android.hardware.display.VirtualDisplay> e
                : sVirtualDisplays.entrySet()) {
            try { e.getValue().release(); } catch (Throwable ignore) {}
        }
        sVirtualDisplays.clear();
    }

    /**
     * Mark a display as single-task instance. BYD/AOSP side-effect: the display
     * becomes "system-trusted" for activity launch, bypassing the secondary-display
     * gate that bounces apps like Waze back to display 0.
     */
    public static String setDisplayToSingleTaskInstance(int displayId) {
        try {
            Class<?> atmCls = Class.forName("android.app.ActivityTaskManager");
            Object iAtm = atmCls.getMethod("getService").invoke(null);
            Method m = iAtm.getClass().getMethod("setDisplayToSingleTaskInstance", int.class);
            m.invoke(iAtm, displayId);
            return "OK setDisplayToSingleTaskInstance(" + displayId + ")";
        } catch (Throwable t) {
            Throwable c = (t instanceof java.lang.reflect.InvocationTargetException
                    && t.getCause() != null) ? t.getCause() : t;
            return "ERR setDisplayToSingleTaskInstance: " + c.getClass().getSimpleName()
                    + " — " + c.getMessage();
        }
    }

    /** Diagnostic: ask ATM if {@code packageName} is allowed to start on {@code displayId}. */
    public static String isActivityStartAllowedOnDisplay(int displayId, String packageName) {
        try {
            Class<?> atmCls = Class.forName("android.app.ActivityTaskManager");
            Object iAtm = atmCls.getMethod("getService").invoke(null);
            android.content.Intent intent = new android.content.Intent(
                    android.content.Intent.ACTION_MAIN);
            intent.addCategory(android.content.Intent.CATEGORY_LAUNCHER);
            intent.setPackage(packageName);
            Method m = iAtm.getClass().getMethod("isActivityStartAllowedOnDisplay",
                    int.class, android.content.Intent.class, String.class, int.class);
            Object res = m.invoke(iAtm, displayId, intent, (String) null, 0);
            return "isActivityStartAllowedOnDisplay(" + displayId + "," + packageName + ") = " + res;
        } catch (Throwable t) {
            Throwable c = (t instanceof java.lang.reflect.InvocationTargetException
                    && t.getCause() != null) ? t.getCause() : t;
            return "ERR isActivityStartAllowedOnDisplay: " + c.getClass().getSimpleName()
                    + " — " + c.getMessage();
        }
    }
}

File: app/src/main/java/com/byd/dashcast/proxy/daemon/Phase4DisplayVerbs.kt
package com.byd.dashcast.proxy.daemon

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.hardware.display.DisplayManager
import android.hardware.display.VirtualDisplay
import android.os.IBinder
import android.view.Display
import android.view.Surface

import com.byd.dashcast.util.concurrent.DeathLease

import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.concurrent.ConcurrentHashMap

/**
 * Phase4DisplayVerbs — display-management verbs that run inside the daemon process (uid 2000).
 *
 * Covers three concern groups:
 *  1. **IWindowManager** — overscan insets via the `window` service.
 *  2. **VirtualDisplay** — create / release VDs on behalf of the app (Phase 5).
 *  3. **ATM display policy** — `setDisplayToSingleTaskInstance`,
 *     `isActivityStartAllowedOnDisplay`.
 *
 * @see Phase4ProcessVerbs
 * @see Phase4TaskVerbs
 * @since v1.1.9 build 174 — split from Phase4Verbs in v1.4.4-beta.
 */
object Phase4DisplayVerbs {

    // ─── IWindowManager cache ─────────────────────────────────────────────

    @Volatile private var sWindowManager: Any? = null
    @Volatile private var sWindowManagerBinder: IBinder? = null
    @Volatile private var sSetOverscan: Method? = null

    private val sWindowManagerDeath = IBinder.DeathRecipient {
        synchronized(Phase4DisplayVerbs::class.java) {
            sWindowManager = null
            sWindowManagerBinder = null
        }
    }

    @Throws(Throwable::class)
    private fun windowManager(): Any? {
        var cached = sWindowManagerBinder
        if (cached != null && cached.isBinderAlive) return sWindowManager
        synchronized(Phase4DisplayVerbs::class.java) {
            cached = sWindowManagerBinder
            val alive = cached
            if (alive != null && alive.isBinderAlive) return sWindowManager
            val sm = Class.forName("android.os.ServiceManager")
            val b = sm.getMethod("getService", String::class.java)
                    .invoke(null, "window") as IBinder?
                ?: throw IllegalStateException("no 'window' service")
            // \$ escape, not a Kotlin template: "IWindowManager$Stub" is a JVM nested-class
            // binary name and "Stub" is a valid Kotlin identifier, so an unescaped version
            // would try to interpolate it.
            val stub = Class.forName("android.view.IWindowManager\$Stub")
            val wm = stub.getMethod("asInterface", IBinder::class.java).invoke(null, b)
                ?: throw IllegalStateException("IWindowManager.asInterface returned null")
            try { b.linkToDeath(sWindowManagerDeath, 0) }
            catch (t: Throwable) { return wm }
            sWindowManagerBinder = b
            sWindowManager = wm
            return wm
        }
    }

    @Throws(Throwable::class)
    private fun setOverscanMethod(): Method {
        var m = sSetOverscan
        if (m != null) return m
        synchronized(Phase4DisplayVerbs::class.java) {
            m = sSetOverscan
            val cached = m
            if (cached != null) return cached
            val iface = Class.forName("android.view.IWindowManager")
            val resolved = iface.getMethod("setOverscan",
                    Int::class.javaPrimitiveType, Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType, Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType)
            sSetOverscan = resolved
            return resolved
        }
    }

    // ─── VirtualDisplay map ───────────────────────────────────────────────

    private val sVirtualDisplays = ConcurrentHashMap<Int, VirtualDisplayRecord>()

    private class VirtualDisplayRecord(display: VirtualDisplay) {
        private var display: VirtualDisplay? = display
        private var ownerLease: DeathLease? = null
        private var released = false

        @Synchronized
        fun setOwnerLease(lease: DeathLease) {
            if (released) lease.close() else ownerLease = lease
        }

        @Synchronized
        fun release() {
            if (released) return
            released = true
            val lease = ownerLease
            ownerLease = null
            lease?.close()
            val current = display
            display = null
            if (current != null) {
                try { current.release() } catch (ignore: Throwable) {}
            }
        }
    }

    // ─── Verbs ────────────────────────────────────────────────────────────

    /**
     * Equivalent of `wm overscan L,T,R,B -d displayId`.
     * Probe P1 (build 173) confirmed viable from uid 2000 on BYD Seal EU.
     */
    @JvmStatic
    @Throws(Throwable::class)
    fun setOverscan(displayId: Int, left: Int, top: Int, right: Int, bottom: Int) {
        val wm = windowManager()
        val m = setOverscanMethod()
        m.invoke(wm, displayId, left, top, right, bottom)
    }

    /**
     * Create a VirtualDisplay on behalf of the calling app and return its display id.
     * The daemon runs as uid 2000 (shell) — uses a `com.android.shell` package context so
     * DisplayManagerService's uid-package check passes.
     *
     * Callers MUST eventually call [releaseVirtualDisplay]; Java GC does NOT release the
     * underlying IVirtualDisplay token.
     */
    @JvmStatic
    fun createVirtualDisplay(sysCtx: Context?,
                             name: String?, width: Int, height: Int, dpi: Int,
                             surface: Surface?, flags: Int,
                             owner: IBinder?): Int {
        if (sysCtx == null) throw IllegalStateException("system context null")
        var vdName = name
        if (vdName == null || vdName.isEmpty()) vdName = "DashCast_VD"
        if (width <= 0 || height <= 0 || dpi <= 0) {
            throw IllegalArgumentException("bad geometry " + width + "x" + height + "@" + dpi)
        }
        if (surface == null || !surface.isValid) {
            throw IllegalArgumentException("surface null or invalid")
        }
        val shellCtx: Context
        try {
            shellCtx = sysCtx.createPackageContext("com.android.shell",
                    Context.CONTEXT_INCLUDE_CODE or Context.CONTEXT_IGNORE_SECURITY)
        } catch (nnfe: PackageManager.NameNotFoundException) {
            throw IllegalStateException("com.android.shell context unavailable", nnfe)
        }
        val dm = shellCtx.getSystemService(Context.DISPLAY_SERVICE) as DisplayManager?
            ?: throw IllegalStateException("DisplayManager null")
        val vd = dm.createVirtualDisplay(vdName, width, height, dpi, surface, flags)
            ?: throw IllegalStateException("createVirtualDisplay returned null")
        val d: Display? = vd.display
        if (d == null) {
            try { vd.release() } catch (ignore: Throwable) {}
            throw IllegalStateException("VirtualDisplay.getDisplay() null")
        }
        val id = d.displayId
        val record = VirtualDisplayRecord(vd)
        val replaced = sVirtualDisplays.put(id, record)
        replaced?.release()
        if (owner != null) {
            try {
                val lease = DeathLease.attach(BinderDeathOwner(owner)) {
                    if (sVirtualDisplays.remove(id, record)) record.release()
                }
                record.setOwnerLease(lease)
            } catch (t: Throwable) {
                sVirtualDisplays.remove(id, record)
                record.release()
                throw IllegalStateException("cannot link VirtualDisplay owner", t)
            }
        }
        return id
    }

    /** Release a VD previously created via [createVirtualDisplay]. No-op if unknown. */
    @JvmStatic
    fun releaseVirtualDisplay(displayId: Int) {
        sVirtualDisplays.remove(displayId)?.release()
    }

    /** Release every VD held by the daemon. Best-effort cleanup. */
    @JvmStatic
    fun releaseAllVirtualDisplays() {
        for (e in sVirtualDisplays.entries) {
            if (sVirtualDisplays.remove(e.key, e.value)) e.value.release()
        }
    }

    /**
     * Mark a display as single-task instance. BYD/AOSP side-effect: the display becomes
     * "system-trusted" for activity launch, bypassing the secondary-display gate that bounces
     * apps like Waze back to display 0.
     */
    @JvmStatic
    fun setDisplayToSingleTaskInstance(displayId: Int): String {
        return try {
            val atmCls = Class.forName("android.app.ActivityTaskManager")
            val iAtm = atmCls.getMethod("getService").invoke(null)
            val m = iAtm!!.javaClass.getMethod("setDisplayToSingleTaskInstance",
                    Int::class.javaPrimitiveType)
            m.invoke(iAtm, displayId)
            "OK setDisplayToSingleTaskInstance($displayId)"
        } catch (t: Throwable) {
            val c = if (t is InvocationTargetException && t.cause != null) t.cause!! else t
            "ERR setDisplayToSingleTaskInstance: " + c.javaClass.simpleName + " — " + c.message
        }
    }

    /** Diagnostic: ask ATM if `packageName` is allowed to start on `displayId`. */
    @JvmStatic
    fun isActivityStartAllowedOnDisplay(displayId: Int, packageName: String?): String {
        return try {
            val atmCls = Class.forName("android.app.ActivityTaskManager")
            val iAtm = atmCls.getMethod("getService").invoke(null)
            val intent = Intent(Intent.ACTION_MAIN)
            intent.addCategory(Intent.CATEGORY_LAUNCHER)
            intent.setPackage(packageName)
            val m = iAtm!!.javaClass.getMethod("isActivityStartAllowedOnDisplay",
                    Int::class.javaPrimitiveType, Intent::class.java, String::class.java,
                    Int::class.javaPrimitiveType)
            val res = m.invoke(iAtm, displayId, intent, null as String?, 0)
            "isActivityStartAllowedOnDisplay($displayId,$packageName) = $res"
        } catch (t: Throwable) {
            val c = if (t is InvocationTargetException && t.cause != null) t.cause!! else t
            "ERR isActivityStartAllowedOnDisplay: " + c.javaClass.simpleName + " — " + c.message
        }
    }
}
